Handle connection to MongoDB with self-signed SSL certificate

gaetanmaisse opened the pull request:
**Issue**

https://github.com/gravitee-io/_support_team/issues/151
https://github.com/gravitee-io/issues/issues/7539

**Description**

 - Add new attributes to handle Truststore in MongoFactory
 - Handle connection to MongoDB with self-signed SSL certificate

**To test**

See section below ⬇️  

----

# APIM + Mongo with SSL 

## Prerequisites

To follow this tutorial, you will need to have the following:
- Docker installed and running
- `openssl` command is available
- `keytool` command is available 

For convenience, you can create a folder to store all the certificates and keys at the same place: 
```shell
mkdir mongo-ssl
cd mongo-ssl
```

## Create the certificates

First, we need to create an SSL certificate, signed by itself, that will be used as root certificate:
```shell
openssl req -newkey rsa:4096 -keyform PEM -keyout ca.key -x509 -days 3650 -subj "/emailAddress=contact@graviteesource.com/CN=mongo.gravitee.io/OU=GraviteeSource/O=GraviteeSource/L=Lille/ST=France/C=FR" -passout pass:ca-secret -outform PEM -out ca.pem
```

It will create a file named `ca.pem` containing the certificate and a file named `ca.key` containing the private key.

Then, we need to create a certificate that we will use on the MongoDB side that needs to be signed by the root certificate:
```shell
# First, create a private key
openssl genrsa -out mongo.key 4096

# Then, create a certificate request
openssl req -new -key mongo.key -out mongo.csr -sha256 -subj "/emailAddress=contact@graviteesource.com/CN=localhost/OU=Mongo/O=GraviteeSource/L=Lille/ST=France/C=FR"

# Finally, sign the certificate request with the root certificate
openssl x509 -req -in mongo.csr -CA ca.pem -CAkey ca.key -set_serial 100 -extensions server -days 1460 -outform PEM -out mongo.cer -sha256 -passin pass:ca-secret

# Create .pem file
cat mongo.key mongo.cer > mongo.pem

# Create a keystore with the root certificate
keytool -import -file ca.pem -storetype PKCS12 -keystore ca-truststore.p12 -storepass truststore-secret -noprompt -alias mongo-ca
```

## Configure and Start MongoDB

Create a `mongod.conf` file with the following content:
```yaml
net:
  bindIp: 0.0.0.0
  port: 27017
  tls:
    certificateKeyFile: /etc/mongo/mongo.pem
    mode: requireTLS
```

Start a MongoDB instance with Docker with the following command:
```shell
# Start MongoDB container with host binding to port 27017
docker run --rm -p 27017:27017 --name mongo-with-ssl -v $(pwd):/etc/mongo mongo:5 --config /etc/mongo/mongod.conf
```

## Configure and start Gravitee APIM

As of 3.15.x, Gravitee APIM can be configured to use Mongo + SSL with a self-signed certificate. 
To do so you need to add the following configuration to the `gravitee.yml` file (both on Management API and Gateway):
```yaml
# In the management.mongodb section
    truststore:
      type: pkcs12                                                              # Supports jks, pem, pkcs12
      path: <<path to your certificate folder>>/mongo-ssl/ca-truststore.p12     # path to KeyStore (when sslEnabled is true, default null)
      password: truststore-secret                                               # KeyStore password (when sslEnabled is true, default null)
```

Et voilà, you are ready to start Gravitee APIM! 🚀
<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-krekrcggxy.chromatic.com)
<!-- Storybook placeholder end -->
<!-- UI placeholder -->
🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/151-mongo-self-signed-certificate/index.html)
_Notes_: The deployed app is linked to the management API of the Element Zero team's environment.
<!-- UI placeholder end -->

gaetanmaisse committed:
fix: handle connection to MongoDB with self-signed SSL certificate

https://github.com/gravitee-io/_support_team/issues/151

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/common/MongoFactory.java:
```diff
@@ -139,35 +140,65 @@ private ServerSettings buildServerSettings() {
+                sslContext.init(getKeyManagers(), getTrustManagers(), null);
+                sslBuilder.context(sslContext);
+            } catch (NoSuchAlgorithmException | KeyManagementException e) {
+                logger.error(e.getCause().toString());
```
The exception is thrown, does the logger.error is useful ?

gaetanmaisse replied:
True, I will clean the other ones too

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-distribution/src/main/resources/config/gravitee.yml:
```diff
@@ -116,7 +116,10 @@ management:
 #    keystorePassword:            # KeyStore password (when sslEnabled is true, default null)
 #    keyPassword:                 # password for recovering keys in the KeyStore (when sslEnabled is true, default null)
-
+#    truststore:
+#      path:                      # Path to the truststore (when sslEnabled is true, default null)
+#      type:                      # Type of the truststore, supports jks, pem, pkcs12 (when sslEnabled is true, default null)
+#      password:                  # Truststore password (when sslEnabled is true, default null)
```
On one side you have `keystore`, `keystorePassword` and no `type` and for the truststore the syntax is `truststore.path`, `truststore.type`, `truststore.passord`. 

I personally prefer the truststore configuration syntax but it is not aligned with the keystore one. What if we could provide the same kind of syntax for the keystore configuration while keeping the old one for backward compatibility ? That way it will be much more convenient and understandable for new joiners. WDYT ?

gaetanmaisse replied:
To be honest, I asked myself the same thing and I totally agree with you.
I was just unsure how to deal with this regarding the doc because making this change is ok (we can make it backward compatible) so the current doc will be still right but if I update the doc then it will contain stuff that is not working with 3.10 😞 .

Otherwise, I need to target 3.10.x 🤔

gaetanmaisse replied:
Should we move forward by pushing these changes and adding a small "info" block in the doc describing keys for 3.10?

jhaeyaert replied:
Maybe it can be done for the 3.10.x also and then the documentation will be valid for both 3.15 and 3.10 (meaning that old `keystoreXXX` config is still working but deprecated in favor of `keystore.path`, `keystore.password`, ....)

gaetanmaisse replied:
Done, I've done the changes and rebased the branch on 3.10.x 🚀  I will update the doc accordingly when this PR will be ✅

jhaeyaert reviewed (commented):
I've let some suggestions (🙂) but LGTM 👍

gaetanmaisse committed:
refactor: rework keystore related attributes used in MongoDB factory

Use the same syntax as the one of `truststore` and add `keystore.type`.
The old properties are still there for backward compatibility.

gaetanmaisse committed:
test: remove unneeded import coming from later version after a backport

Fix unit tests on 3.10.x after the merge of https://github.com/gravitee-io/gravitee-api-management/pull/1616
